lua/telescope/health.lua

```diff
@@ -0,0 +1,133 @@
+        optional = false,
+      },
+    },
+  },
+  {
+    finder_name = "find-files",
+    package = {
+      {
+        name = "fd",
+        url = "[sharkdp/fd](https://github.com/sharkdp/fd)",
+        optional = true,
+      },
+    },
+  },
+}
+
+local required_plugins = {
+  { lib = "popup", optional = false },
+  { lib = "plenary", optional = false },
+  {
+    lib = "nvim-treesitter",
+    optional = true,
+    info = "",
+  },
+}
+
+local check_binary_installed = function(package)
+  local file_extension = is_win and ".exe" or ""
+  local filename = package.name .. file_extension
+  if fn.executable(filename) == 0 then
+    return
+  else
+    local handle = io.popen(filename .. " --version")
+    local binary_version = handle:read "*a"
+    handle:close()
+    return true, binary_version
+  end
+end
+
+local function lualib_installed(lib_name)
+  local res, _ = pcall(require, lib_name)
+  return res
+end
+
+-- local function get_healthcheck(ext_name)
+--   local has_check, check = pcall(require, "telescope.extensions"[ext_name]['health'])
+--     if(has_check) then return check end
+--   return
+-- end
+
+local M = {}
+
+M.check_health = function()
+  -- Required lua libs
+  health_start "Checking for required plugins"
+  for _, plugin in ipairs(required_plugins) do
+    if lualib_installed(plugin.lib) then
+      health_ok(plugin.lib .. " installed.")
+    else
+      local lib_not_installed = plugin.lib .. " not found."
+      if plugin.optional then
+        health_warn(("%s %s"):format(lib_not_installed, plugin.info))
+      else
+        health_error(lib_not_installed)
+      end
+    end
+  end
+
+  -- external dependencies
+  -- TODO: only perform checks if user has enabled dependency in their config
+  health_start "Checking external dependencies"
+
+  for _, opt_dep in pairs(optional_dependencies) do
+    for _, package in ipairs(opt_dep.package) do
+      local installed, version = check_binary_installed(package)
+      if not installed then
+        local err_msg = ("%s: not found."):format(package.name)
+        if package.optional then
+          health_warn(("%s %s"):format(err_msg, ("Install %s for extended capabilities"):format(package.url)))
+        else
+          health_error(
+            ("%s %s"):format(
+              err_msg,
+              ("`%s` finder will not function without %s installed."):format(opt_dep.finder_name, package.url)
+            )
+          )
+        end
+      else
+        local eol = version:find "\n"
+        health_ok(("%s: found %s"):format(package.name, version:sub(0, eol - 1) or "(unknown version)"))
+      end
+    end
+  end
+
+  -- Extensions
+  health_start "Installed extensions"
+  local installed = {}
+  for extension_name, _ in pairs(extension_info) do
+    installed[#installed + 1] = extension_name
+  end
+  table.sort(installed)
+
+  for _, installed_ext in ipairs(installed) do
+    local extension_healthcheck = extension_info[installed_ext].health
```

Comment:
sorry i am still kinda busy. The "issue" that i have is that health is now exported and will appear when doing. `Telescope frecency health`.

On the one hand its not the worst but on the other i am not sure i really like it. I had more something like this in mind (from extension side).

```lua
return require("telescope").register_extension {
  setup = function(ext_config, config)
    -- config
  end,
  exports = {
    -- the important public thins
  },
  health = function()
   local health_ok = vim.fn["health#report_ok"]
   health_ok "okay"
  end
}
```

for implementation we currently have in `_extensions` a `manager` table which holds all extensions. For health, we could just write that in a separate local table under extensions. Similar to `_loaded` or `_config`. Like `extension._health[name] = fn` and then you could all that here.

Does this make sense? Or am i overthinking this :rofl:

Reply:
Yes, adding a `health` argument for `register_extension` is very nice. I prefer this.